src/UserDataContext.tsx

```diff
@@ -129,6 +135,8 @@ export const UserDataContextProvider: React.FC<{
     i18n.changeLanguage(newlanguage.toString());
   };
 
+  const logout = () => clearUserData(ctx);
```

Comment:
This is "hoisting" for variables, right? I've never seen it before. So just for me to understand, this code is the same as:
```ts
const ctx = {
  // ...
  logout: () => {},
};
ctx.logout = () => clearUserData(ctx);
```
Correct?

Reply:
I think youre right.

i think this uses hoisting, as things arent declared in the right order, kind of.
https://developer.mozilla.org/en-US/docs/Glossary/Hoisting
to avoid what seems like a circular dependency, i could pass only the functions used in the `clearUserData` (`deleteUserDataFromLS`,`setUserID`,`setUserName`,`setAuthenticationToken`). this way, logout can be fully declared before being part of the same context that it uses. 
i didnt notice this before you pointed it out.
but JS/React seems to be handling this fine 🤷 it might break if `logout` called `ctx.logout`

(I think you know:) This follows the same pattern as the `makeNotifyUserOnNotLoggedInError` function in the same file, except it is exposed through the context values.
The reason to do it this way is to `bind` the function to the current context. https://javascript.info/bind
so essentially scope management